File: Algorithms/041. First Missing Positive/Solution.py
#coding=utf-8
__author__ = 'xuxuan'
class Solution(object):
    def firstMissingPositive(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        tar=0
        nums.append(0)
        l=len(nums)
        for i in range(l):
            b=i
            a=nums[i]
            while a>0 and a<l and a!=b:
                tar=nums[a]
                nums[a]=a
                b=a
                a=tar
        for i in range(1,len(nums)):
            if nums[i]!=i:
                return i
        return len(nums)

File: Algorithms/041. First Missing Positive/Solution.java
public class Solution {
    public int firstMissingPositive(int[] nums) {
        for (int i = 0; i < nums.length; i++) {
            if (nums[i]!=i+1) {
                if (nums[i]<=nums.length&&nums[i]>0&&nums[nums[i]-1]!=nums[i]) {
                    int k=nums[nums[i]-1];
                    nums[nums[i]-1]=nums[i];
                    nums[i]=k;
                    i--;
                }
            }
        }
        for (int i = 0; i < nums.length; i++) {
            if (nums[i]!=i+1) {
                return i+1;
            }
        }
        return nums.length+1;
    
        
    }
}